src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,316 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021-2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/surf/detail/QuadraticSolver.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cmath>
+#include <iostream>
+#include <vector>
+
+#include "corecel/Types.hh"
+#include "corecel/cont/Array.hh"
+#include "corecel/math/Algorithms.hh"
+#include "orange/OrangeTypes.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Find positive, real, nonzero roots for involute intersection function.
+ *
+ * The involute intersection equation \f[
+   r_b * [v{cos(t+a)+tsin(t+a)} + u{sin(t+a)-tcos(t+a)}] + xv - yu = 0
+ * \f]
+ * has n solutions mathematically, but we only want solutions where t is
+ results
+ * in a real, positive, and in the defined bounds.  Furthermore the equation is
+ * subject to catastrophic roundoff due to floating point precision (see
+ * \c Tolerance::sqrt_quadratic and the derivation in \c CylAligned ).
+ *
+ * \return An Intersections array where each item is a positive valid
+ * intersection or the sentinel result \c no_intersection() .
+ */
+class InvoluteSolver
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Intersections = Array<real_type, 3>;
+
+  public:
+    // Construct with
+    inline CELER_FUNCTION InvoluteSolver(real_type r_b_,
+                                         real_type a_,
+                                         real_type sign_,
+                                         real_type tmin_,
+                                         real_type tmax_);
+
+    // Solve fully general case
+    inline CELER_FUNCTION Intersections operator()(real_type x,
+                                                   real_type y,
+                                                   real_type z,
+                                                   real_type u,
+                                                   real_type v,
+                                                   real_type w) const;
```

Comment:
Use `Real3 const& pos` and `Real3 const& dir` if possible; it makes the function less error-prone since it logically separates the two variables rather than flattening them together.